contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,338 @@
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
+        game.playerTwo = msg.sender;
+        game.playerTwoMoveHash = moveHash;
+        game.step = uint8(Steps.PLAYER_TWO_MOVE);
+        game.expiryDate = expiryDate;
+        depositStake(game.stake);
+
+        emit PlayerTwoMoves(msg.sender, msg.value, gameKey, moveHash, expiryDate);
+    }
+
+    function depositStake(uint256 stake) internal {
+
+        if (msg.value == stake) {
+            return;
+        }
+
+        uint256 senderBalance = balances[msg.sender];
+        if (msg.value < stake) {
+            require(senderBalance >= stake.sub(msg.value), "Insufficient balance");
+        }
+
+        balances[msg.sender] = senderBalance.add(msg.value).sub(stake);
+    }
+
+    function revealPlayerOne(bytes32 secret, uint256 playerOneMove) external whenNotPaused {
+
+        require(playerOneMove > 0 && playerOneMove < 4, INVALID_MOVE_MSG);
+        bytes32 gameKey = createPlayerOneMoveHash(msg.sender, secret, playerOneMove);
+        Game storage game = games[gameKey];
+        require(game.step == uint8(Steps.PLAYER_TWO_MOVE), INVALID_STEP_MSG);
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
+        game.playerOneMove = uint8(playerOneMove);
+        game.step = uint8(Steps.PLAYER_ONE_REVEAL);
+        game.expiryDate = expiryDate;
+
+        emit PlayerOneReveals(msg.sender, gameKey, uint8(playerOneMove), expiryDate);
+    }
+
+    function revealPlayerTwo(bytes32 gameKey, bytes32 secret, uint256 playerTwoMove) external whenNotPaused {
+
+        require(playerTwoMove > 0 && playerTwoMove < 4, INVALID_MOVE_MSG);
+        Game storage game = games[gameKey];
+        require(game.step == uint8(Steps.PLAYER_ONE_REVEAL), INVALID_STEP_MSG);
+        address playerTwo = game.playerTwo;
+
+        //Validate move
+        bytes32 expectedMoveHash = createPlayerTwoMoveHash(msg.sender, gameKey, secret, playerTwoMove);
+        require(game.playerTwoMoveHash == expectedMoveHash, HASH_MISMATCH_MSG);
+
+        emit PlayerTwoReveals(msg.sender, gameKey, uint8(playerTwoMove));
+
+        uint256 stake = game.stake;
+        address playerOne = game.playerOne;
+        uint256 playerOneMove = uint256(game.playerOneMove);
+        address winner;
+        address loser;
+        uint256 outcome = resolveGame(playerOneMove, playerTwoMove);
+
+        if (outcome == uint256(Moves.DRAW)) {
+            resetGame(game);
+            balances[playerOne] = balances[playerOne].add(stake);
+            balances[playerTwo] = balances[playerTwo].add(stake);
+            DrawRefund(playerOne, playerTwo, gameKey, stake, stake);
+
+            return;
+
+        } else if (outcome == uint256(Moves.LEFTWIN)) {
+            winner = playerOne;
+            loser = playerTwo;
+        } else if (outcome == uint256(Moves.RIGHTWIN)) {
+            winner = playerTwo;
+            loser = playerOne;
+        } else {
+            revert(INVALID_MOVE_MSG);
```

Comment:
Is a revert suitable? I am thinking this because `assert` has no error message.

Reply:
Well, what good is an error message for a line that will never be reached, right?

Reply:
If it will never get here then can I not simplify to:
```
        if (outcome == GameResult.DRAW) {
            resetGame(game);
            balances[playerOne] = balances[playerOne].add(stake);
            balances[playerTwo] = balances[playerTwo].add(stake);
            DrawRefund(gameKey, playerOne, playerTwo, stake, stake);

            return;

        } else if (outcome == GameResult.LEFTWIN) {
            winner = playerOne;
            loser = playerTwo;
        } else {
            winner = playerTwo;
            loser = playerOne;
        }
```

Reply:
That is also one way to play it. Although, debatably, for future formal verification, I would add:

```solidity
else {
    assert(outcome == GameResult.RIGHTWIN);
...
```

Reply:
OK. Thanks. That is interesting. I have been writing if else statements for years and never satisfactorily resolved how to best handle this situation. Probably I would have thrown the equivalent of an UnexpectedValue exception.